### ISSUE TITLE
Delegate support for matching spied arguments

### ISSUE THREAD

**Issue by Nysosis** [alsatian-test/alsatian#304]
Right now we only have the capability to support `Any` or `Any(Type)` on `spy`'d methods, it would be nice if we could support a delegate check on the arguments spied to check if they were valid or not.

**Comment by jamesrichford** [comment]
Hey @Nysosis. Could you give an example of how this would be used please?

**Comment by Nysosis** [comment]
Yup sure, so my actual use case is working with AWS's Lamda, and you need to export a method with the following signature:

`handler(event: any, context: any, callback: (error?: any, result?: any) => void)`

With the `callback` param being passed the output of the work, or an error if it failed (not the nicest method, I'd rather use promises, but that's beside the point. So if I want to test my method, the only way I have access to the output is through that callback method.

I know that I could just check the parameters as the method is called:

```
handler(event, context, (err, res) => {
    Expect(err).toBe(null);
});
```

But I like to stick close to the `Act/Assert/Arrange` format for writing tests, so this change (in #305) would allow me to write the test more like:

```
const spy = new FunctionSpy();

handler(null, null, spy.call);

Expect(spy).toHaveBeenCalledWith(Matches(Error, e => e.message === "My expected error")).exactly(1).times;
```

Or if you've got a class which is taking one input, transforming it, and passing to a method on an injected interface, you might want to check the transform is correct by validating the objects passed to the dependency method.

**Comment by jamesrichford** [comment]
OK so this looks like it may be linked to #221 if I understand correctly?

In that I want to do a type check but then additional checks

```typescript
Any(Error).with(e => e.message === "My expected error")

// or perhaps in the case of a function which?
Any(Error).which(e => e.message === "My expected error")
```

This syntax should also support the neater syntax when you are doing a simple comparison

```typescript

Any(Error).with("message", "My expected error")

// or

Any(Error).which({ message: "My expected error" })
```

Are the two similar in any way or have I misinterpreted the request?

**Comment by Nysosis** [comment]
That does seem to be the same sort of idea yes, just different syntax I
guess! (Mine was based off moq <https://github.com/Moq/moq4/wiki/Quickstart>'s
`It.Is` syntax, which is what I'm primarily used to)

**Comment by jamesrichford** [comment]
Ah cool, maybe it's with bring forward the 1.3 milestone then. The focus of the milestone will be on spying. Is there anything else to do with spying that would be useful to improve.

Also how would you feel about the proposed syntax to solve this solution using an any with style?

**Comment by Nysosis** [comment]
I like the proposed syntax, reads really easy, I'd go for `with` for property matching, and `which` for a delegate.

I can't say I've ever had much of a need outside being able to check equivalency of an argument, either through property assertion or delegate.

**Comment by jamesrichford** [comment]
I guess my only thought about the proposed syntax is `with` and `which` are very close, maybe change `with` to `that` to avoid confusion. Particularly for someone who's first language isn't English. What do you think? Of course if you can think of a better word that'd be much appreciated :)

**Comment by Jameskmonger** [comment]
I like the FakeItEasy syntax:

**Check for `hello` and any `int`**

```
A.CallTo(() => foo.Bar("hello", A<int>.Ignored)).MustHaveHappened();
```

**Check for `hello` and any `int` between 200 and 40.

```
A.CallTo(() => foo.Bar("hello", 
    A<int>.That.Matches(i => i => 20 && i <= 40)
).MustHaveHappened();
```

**Comment by Nysosis** [comment]
Similar to Moq's

*Check for hello and any int*

`myMock.Verify(foo => foo.Bar("hello", It.IsAny<int>(), Times.Once());`

*Check for hello and any int between 20 and 40.*

`myMock.Verify(foo => foo.Bar("hello", It.Is<int>(i => i >= 20 && i <= 40),
Times.Once());`

(Was never a fan of the `Times.Once()` though)

I think it will always come down to personal preference of a developer how
they'd rather read/write the statement, as there's many equally valid ways
of writing it out. Just as long as whatever one is chosen for the spying
makes sense then I don't think it overly matters.

**Comment by jamesrichford** [comment]
OK so when making a simple comparison we'll use

```typescript
Expect(a.function).toHaveBeenCalledWith(Any(Object).with("name", "James"));
```

and for those that take a delegate

```typescript
Expect(a.function).toHaveBeenCalledWith(Any(number).thatMatches(n => n >= 20 && n <= 40));
```

All sound good @Nysosis & @Jameskmonger?

**Comment by Jameskmonger** [comment]
So in the first example (`Any(Object).with("name", "James")`) it would expect `{ name: "James" }`?

**Comment by jamesrichford** [comment]
Here's the behaviour

```typescript
Expect(a.function).toHaveBeenCalledWith(Any(Object).with("name", "James"));

a.function({ name: "James" }); // pass
a.function({ name: "James", another: "property" }); // pass
a.function({ name: "J" }); // fail
a.function({ another: "property" }); // fail
a.function({  }); // fail
```

Could also have an overload to allow multiple properties
```
Expect(a.function).toHaveBeenCalledWith(Any(Object).with({
   name: "James",
   awesomeness: 1000000
}));
```

**Comment by Jameskmonger** [comment]
I'm not entirely sure there's a need for the `with` at all and I think it could probably be a bit confusing.

The purpose of the `thatMatches` function is that you can check that the parameter meets some loose criteria - is a number above 10, is the string longer than 5, etc.

As you showed in your example, checking that `name` is `James` still passes with other properties, as long as the `name` is correct - I would say that this is a perfect use case for the `thatMatches` function also, as you just want to check some properties of the object and not necessarily all of them:

    Expect(a.function).toHaveBeenCalledWith(Any(Object).thatMatches(o => o.name === "James"));

    a.function({ name: "James" }); // pass
    a.function({ name: "James", another: "property" }); // pass
    a.function({ name: "J" }); // fail
    a.function({ another: "property" }); // fail
    a.function({  }); // fail

**Comment by jamesrichford** [comment]
Yep for sure there's no need for it :) but it's just adding a neater way of writing the simpler check

```typescripts
Expect([1, 2, 3].indexOf(2)).not.toBe(-1);
Expect([1, 2, 3]).toContain(2);
```

Both the above are the same but one reads a bit more clear maybe to avoid confusion these functions should all be `thatMatches` overloads so you can pass a function, object or two arguments?

**Comment by Jameskmonger** [comment]
I think that having them all as overloads would be a good idea. What do you think @Nysosis?

**Comment by Nysosis** [comment]
Yup, multiple overloads sounds good to me!

**Comment by jamesrichford** [comment]
closed by #305
